fetching/pick-fetcher/src/index.ts

```diff
@@ -1,23 +1,67 @@
 import type { AtomicResolution } from '@pnpm/resolver-base'
-import type { Fetchers, FetchFunction, DirectoryFetcher, GitFetcher, BinaryFetcher } from '@pnpm/fetcher-base'
+import type { Fetchers, FetchFunction, DirectoryFetcher, GitFetcher, BinaryFetcher, FetchOptions } from '@pnpm/fetcher-base'
+import type { Cafs } from '@pnpm/cafs-types'
+import { PnpmError } from '@pnpm/error'
+import { type CustomFetcher } from '@pnpm/hooks.types'
 
-export function pickFetcher (fetcherByHostingType: Partial<Fetchers>, resolution: AtomicResolution): FetchFunction | DirectoryFetcher | GitFetcher | BinaryFetcher {
-  let fetcherType: keyof Fetchers | undefined = resolution.type
+export async function pickFetcher (
+  fetcherByHostingType: Fetchers,
+  resolution: AtomicResolution,
+  opts?: {
+    customFetchers?: CustomFetcher[]
+    packageId?: string
```

Comment:
why is this optional? Don't we always have a packageId?

Reply:
Fixed by 21c9a211ceb673e145b97e42c3d5a9fe17fca457.